libs/token/src/src_20.sw

```diff
@@ -0,0 +1,341 @@
+library;
+
+use std::{
+    hash::sha256,
+    storage::storage_string::*,
+    string::String,
+};
+
+/// Returns the total number of individual assets for a contract.
+///
+/// # Arguments
+///
+/// * `total_assets_key`: [StorageKey<u64>] - The location in storage that the `u64` which represents the total assets is stored.
+///
+/// # Returns
+///
+/// * [u64] - The number of assets that this contract has minted.
+///
+/// # Number of Storage Accesses
+///
+/// * Reads: `1`
+///
+/// # Examples
+///
+/// ```sway
+/// use token::_total_assets;
+///
+/// storage {
+///     total_assets: u64 = 0,
+/// }
+///
+/// fn foo() {
+///     let assets = _total_assets(storage.total_assets);
+///     assert(assets == 0);
+/// }
+/// ```
+#[storage(read)]
+pub fn _total_assets(total_assets_key: StorageKey<u64>) -> u64 {
+    total_assets_key.try_read().unwrap_or(0)
+}
+
+/// Returns the total supply of tokens for an asset.
+///
+/// # Arguments
+///
+/// * `total_supply_key`: [StorageKey<StorageMap<AssetId, u64>>] - The location in storage which the `StorageMap` that stores the total supply of assets is stored.
+/// * `asset`: [AssetId] - The asset of which to query the total supply.
+///
+/// # Returns
+///
+/// * [Option<u64>] - The total supply of an `asset`.
+///
+/// # Number of Storage Accesses
+///
+/// * Reads: `1`
+///
+/// # Examples
+///
+/// ```sway
+/// use token::_total_supply;
+///
+/// storage {
+///     total_supply: StorageMap<AssetId, u64> = StorageMap {},
+/// }
+///
+/// fn foo(asset_id: AssetId) {
+///     let supply = _total_supply(storage.total_supply, asset_id);
+///     assert(supply.unwrap_or(0) != 0);
+/// }
+/// ```
+#[storage(read)]
+pub fn _total_supply(
+    total_supply_key: StorageKey<StorageMap<AssetId, u64>>,
+    asset: AssetId,
+) -> Option<u64> {
+    total_supply_key.get(asset).try_read()
+}
+
+/// Returns the name of the asset, such as “Ether”.
+///
+/// # Arguments
+///
+/// * `name_key`: [StorageKey<StorageMap<AssetId, StorageKey<StorageString>>>] - The location in storage which the `StorageMap` that stores the names of assets is stored.
+/// * `asset`: [AssetId] - The asset of which to query the name.
+///
+/// # Returns
+///
+/// * [Option<String>] - The name of `asset`.
+///
+/// # Number of Storage Accesses
+///
+/// * Reads: `1`
+///
+/// # Examples
+///
+/// ```sway
+/// use token::_name;
+/// use std::string::String;
+///
+/// storage {
+///     name: StorageMap<AssetId, StorageKey<StorageString>> = StorageMap {},
+/// }
+///
+/// fn foo(asset: AssetId) {
+///     let name = _name(storage.name, asset);
+///     assert(name.unwrap_or(String::new()).len() != 0);
+/// }
+/// ```
+#[storage(read)]
+pub fn _name(
+    name_key: StorageKey<StorageMap<AssetId, StorageKey<StorageString>>>,
+    asset: AssetId,
+) -> Option<String> {
+    match name_key.get(asset).try_read() {
+        Option::Some(s) => s.read_slice(),
+        Option::None(s) => Option::None,
```

Comment:
```suggestion
        Option::None => Option::None,
```

Pretty sure this should be considered a bug

Reply:
This should be addressed in a different PR as it's just the same code copy and pasted into a new file.

Reply:
I really do not follow your reasoning here.
"Copy and paste is okay but adjusting minor syntax is not". If this causes actually issues in tests etc. then yh, another PR otherwise changing is fine